src/binary/writer.rs

```diff
@@ -0,0 +1,1324 @@
+// buffer : e0 01 00 ea | 83 66 6f 6f | 83 62 61 72 | b8 | 81 | 81 | eb
+//          ^             ^             ^             ^    ^    ^    ^-- 11-byte annotation wrapper
+//          |             |             |             |    |    +------- Annotation seq length 1
+//          |             |             |             |    +------------ Annotation 1 ('$ion')
+//          |             |             |             +----------------- 8-byte list header
+//          |             |             +------------------------------- 3-byte string "bar"
+//          |             +--------------------------------------------- 3-byte string "foo"
+//          +----------------------------------------------------------- Ion 1.0 version marker
+//
+// Meanwhile, the writer's IoRange queue would look like this:
+//
+//     0..4, 15..16, 14..15, 13..14, 12..13, 4..12
+//
+// Ranges have inclusive starts and exclusive ends, so the range 14..15 includes a single byte at
+// index 14.
+//
+// When the writer's `flush()` method is called, each IoRange is turned into a slice of the buffer
+// and written to the io::Write sink.
+//
+//     e0 01 00 ea eb 81 81 b8 83 66 6f 6f 83 62 61 72
+
+// Represents a level into which the writer has stepped.
+// A writer that has not yet called step_in() is at the top level.
+#[derive(Debug, PartialEq)]
+enum ContainerType {
+    TopLevel,
+    SExpression,
+    List,
+    Struct,
+}
+
+// Stores information about each level into which the writer has stepped, including its
+// annotations, field_id, and container type.
+#[derive(Debug)]
+struct EncodingLevel {
+    container_type: ContainerType,
+    field_id: Option<SymbolId>,
+    // Annotations are stored in a common Vec on the BinarySystemWriter. Each EncodingLevel tracks
+    // how many annotations it had, allowing that Vec to be treated as a stack. Stepping into
+    // a new level pushes `num_annotations` symbol IDs onto the Vec and stepping out pops
+    // `num_annotations` back off the stack.
+    num_annotations: u8,
+    // Index of the IoRange for this container's type descriptor and length. When the writer
+    // steps out of this level, the type descriptor IoRange will be retrieved and populated.
+    td_io_range_index: usize,
+}
+
+impl EncodingLevel {
+    fn new(container_type: ContainerType,
+           field_id: Option<SymbolId>,
+           num_annotations: u8,
+           td_io_range_index: usize) -> EncodingLevel {
+        EncodingLevel {
+            container_type,
+            field_id,
+            num_annotations,
+            td_io_range_index,
+        }
+    }
+
+    // Visits all of the IoRanges belonging to this EncodingLevel and notes their total length.
+    // This length will be written out as a length prefix for the container.
+    fn calculate_final_size(&self, io_ranges: &mut [Range<usize>]) -> usize {
+        io_ranges[self.td_io_range_index..]
+            .iter()
+            .map(|r| r.len())
+            .sum()
+    }
+}
```

Comment:
I am missing something here, or rather, I am not sure I get the whole picture. The input `io_ranges` and the stored `td_io_range.index` are `usize` based ranges. How do we know when iterating over `io_range` that we are not going over values in the range that are *not* for our container but some other container? I'll keep reading to see if I can figure out the answer, but I am assuming when this function is called the tail end of `io_ranges` starting at `td_io_range_index` contains the information for this container which is also the last "processed" container.

I am assuming an explicit contract here that this method is called at the end of a container which would guarantee that the tail range in `io_ranges` contains the relevant information and nothing extra.

Reply:
>  I am assuming when this function is called the tail end of io_ranges starting at td_io_range_index contains the information for this container which is also the last "processed" container.

This is correct. 

When the writer steps into a container, it pushes a number of placeholder `IoRange`s that will need to be populated later, when more length information is available. Afterwards, as the writer continues writing new nested values inside the container (potentially including other containers), all of their `IoRange`s will be pushed *after* the placeholders we already added. When `step_out()` is called, we look at the `EncodingLevel` that's popped off the stack. By definition, all of the `IoRange`s that follow its type descriptor `IoRange` (at offset `self.td_io_range_index`) belong to nested values.

> I am assuming an explicit contract here that this method is called at the end of a container which would guarantee that the tail range in io_ranges contains the relevant information and nothing extra.

Yep! `calculate_final_size()` is only called during `step_out()`. At the moment a streaming writer is stepping out of a container, there cannot be any values after the container yet.